rpc/send.go

```diff
@@ -261,3 +272,33 @@ func sendOne(client *Client, timeout time.Duration, method string,
 		}
 	}()
 }
+
+// localCall invokes the specified method directly.
+func localCall(s *Server, method string, args proto.Message, done chan *rpc.Call) {
+	s.mu.RLock()
+	m := s.methods[method]
+	s.mu.RUnlock()
+
+	if m.handler == nil {
+		done <- &rpc.Call{
+			Error: newRPCError(util.Errorf("rpc:couldn't find method: %s", method)),
+		}
+		return
+	}
+
+	if _, err := security.CheckRequestUser(args, m.public); err != nil {
+		done <- &rpc.Call{
+			Error: err,
+		}
+		return
+	}
+
+	m.handler(args, func(reply proto.Message, err error) {
```

Comment:
I believe calling the handler synchronously here can deadlock in `server/raft_transport.go`, as I described in #3209. Now that I think about it, #4039 introduced that same deadlock risk to the main code path because `client.Go` is only partially asynchronous (#3250).